src/HPLTokenHandler.mo

```diff
@@ -0,0 +1,331 @@
+
+    /// query journal for debug purposes. Returns:
+    /// 1) array of all items in order, starting from the oldest record in journal, but no earlier than "startFrom" if provided
+    /// 2) the index of next upcoming journal log. Use this value as "startFrom" in your next journal query to fetch next entries
+    public func queryJournal(startFrom: ?Nat): ([JournalRecord], Nat) = (
+      Iter.toArray(
+        journal.slice(
+          Int.abs(Int.max(Option.get(startFrom, 0), journal.pushesAmount() - journalSize)),
+          journal.pushesAmount()
+        )
+      ),
+      journal.pushesAmount(),
+    );
+
+    /// retrieve the current freeze state
+    public func isFrozen() : Bool = isFrozen_;
+
+    /// retrieve the sum of all successful consolidations
+    public func consolidatedFunds() : Nat = consolidatedFunds_;
+
+    /// deduct amount from P’s usable balance. Return false if the balance is insufficient.
+    public func debit(p : Principal, amount : Nat) : Bool {
+      if (isFrozen()) {
+        return false;
+      };
+      let ?info = map.get(p) else return false;
+      if (info.credit < amount) return false;
+      info.credit -= amount;
+      journal.push((Time.now(), p, #debited(amount)));
+      totalDebited += amount;
+      assertBalancesIntegrity();
+      true;
+    };
+
+    ///  add amount to P’s usable balance (the credit is created out of thin air)
+    public func credit(p : Principal, amount : Nat) : Bool {
+      if (isFrozen()) {
+        return false;
+      };
+      let ?info = map.get(p) else return false;
+      info.credit += amount;
+      totalCredited += amount;
+      journal.push((Time.now(), p, #credited(amount)));
+      assertBalancesIntegrity();
+      true;
+    };
+
+    /// The handler will query the balance of virtual account, previously opened for P.
+    /// If there will be non-zero amount (deposit), the handler will turn balance to zero and add the deposit to the credit of P. 
+    /// Returns the newly detected deposit and total usable balance if success, otherwise null
+    public func notify(p : Principal) : async* ?(Nat, Nat) {
+      if (isFrozen()) return null;
+      let ?info = map.get(p) else return null;
+      let updateResult = await hpl.setVirtualBalance(info.virtualAccountId, 0);
+      switch (updateResult) {
+        case (#ok delta) {
+          if (delta == 0) {
+            return ?(0, info.credit);
+          };
+          let deposit = Int.abs(delta); // expected delta here is always negative
+          info.credit += deposit; 
+          journal.push((Time.now(), p, #newDeposit(deposit)));
+          consolidatedFunds_ += deposit;
+          ?(deposit, info.credit);
+        };
+        case (#err err) {
+          let message = "Unexpected error during setting virtual account balance";
+          journal.push((Time.now(), ownPrincipal, #error(message, err)));
+          freezeTokenHandler(message);
```

Comment:
Need to double-check what the possible error causes are and if they require to freeze.

Reply:
Potential errors, that can be returned by HPL:
`#UnknownPrincipal` - cannot happen, unless parent application did not register HPL account for itself e.g. it has a bug. In this case parent application could not provide correct backingSubaccountId, required in handler constructor
`#UnknownVirtualAccount` - cannot happen, because notify gets virtual account id from the map, and it can be added to the map only after we registered virtual account  
`#DeletedVirtualAccount;` - cannot happen, we never delete virtual accounts in the token handler.

Reply:
Ok, then I think freezing and throwing is the desired behaviour.